aten/src/ATen/cpu/vec256/vsx/vsx_helpers.h

```diff
@@ -0,0 +1,326 @@
+#pragma once
+#include <ATen/cpu/vec256/intrinsics.h>
+#include <cstdint>
+
+#define __inline_attrs  __attribute__((__always_inline__))
+using __vd = __vector double;
+using __vf = __vector float;
+using __vchar = __vector unsigned char;
+using __vchari = __vector signed char;
+using __vcharb = __vector bool char;
+using __vllb = __vector bool long long;
+using  __vlli = __vector signed long long;
+using  __vulli = __vector unsigned long long;
+using __vib = __vector bool int;
+using __vui = __vector unsigned  int;
+using __vi = __vector signed int;
+using __vshi = __vector signed short;
+using __vsh = __vector unsigned short;
+using __vshb = __vector bool short;
+
+
+#define DEFINE_MEMBER_OP(op, op_type, func)                              \
```

Comment:
Could you simplify this part by not using macro? The macros in PyTorch have been headache for maintenance. Perhaps something like the following? (not tested, only a rough framework, feel free to adjust it)

```c++
template <typename Op, typename Func>
Vec256<Op> inline __inline_attr op_template(const Vec256<Op>& other, Func func) {
    return Vec256<Op>{func(_vec0, other._vec0), func(_vec1, other._vec1)};
}
```

When using it:

```c++
Vec256<int64_t> __inline_attr operator +(const Vec256<int64_t>& other) {
    return op_template(other, std::plus<int64_t>());
}
```

Reply:
tbh, macrosses way is cleaner as I can exploit overloaded power vector intrinsics  otherwise ordinary way more cleaner than template one

Reply:
I kept the current

Reply:
> Could you simplify this part by not using macro? The macros in PyTorch have been headache for maintenance. Perhaps something like the following? (not tested, only a rough framework, feel free to adjust it)
> 
> ```c++
> template <typename Op, typename Func>
> Vec256<Op> inline __inline_attr op_template(const Vec256<Op>& other, Func func) {
>     return Vec256<Op>{func(_vec0, other._vec0), func(_vec1, other._vec1)};
> }
> ```
> 
> When using it:
> 
> ```c++
> Vec256<int64_t> __inline_attr operator +(const Vec256<int64_t>& other) {
>     return op_template(other, std::plus<int64_t>());
> }
> ```

I will re-check it again.    I will see how one base template class with basic  operations will behave.

Reply:
I postponed it for now.
I wanted to do this way exploiting overloaded vsx functions
```
template<typename T>
struct VsxBaseVec{
//....
// vector_type<T> vec0;
// vector_type<T> vec1;
//...
VsxBaseVec  operator+(const VsxBaseVec &other){
   return { vec_add( _vec0, other._vec0), vec_add( _vec0, other._vec0)};
}
...
};
```

Reply:
I'm OK with postponing it for now. Good work!

Reply:
> I postponed it for now.
> I wanted to do this way exploiting overloaded vsx functions
> 
> ```
> template<typename T>
> struct VsxBaseVec{
> //....
> // vector_type<T> vec0;
> // vector_type<T> vec1;
> //...
> VsxBaseVec  operator+(const VsxBaseVec &other){
>    return { vec_add( _vec0, other._vec0), vec_add( _vec1, other._vec1)};
> }
> ...
> };
> ```

@Flamefire